File: 0211-design-add-and-search-words-data-structure/0211-design-add-and-search-words-data-structure.java
class WordDictionary {

    private Set<String> dic;
    private Deque<String> queue;

    public WordDictionary() {
        this.dic = new HashSet<>();
        this.queue = new ArrayDeque<>();
    }
    
    public void addWord(String word) {
        dic.add(word);
    }
    
    public boolean search(String word) {
        queue.clear();
        queue.offer(word);
        while(!queue.isEmpty()) {
            String cur = queue.poll();
            int dotIndex = cur.indexOf(".");
            if(dotIndex == -1) {
                if(dic.contains(cur)) return true;
            } else {
                String replace;
                for(int i = 0; i < 26; i++) {
                    char dotReplace = (char) ('a' + i);
                    replace = cur.substring(0, dotIndex) + dotReplace + cur.substring(dotIndex+1, cur.length());
                    queue.offer(replace);
                }
                 
            }
        }


        return false;
    }
}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary obj = new WordDictionary();
 * obj.addWord(word);
 * boolean param_2 = obj.search(word);
 */

File: 0211-design-add-and-search-words-data-structure/0211-design-add-and-search-words-data-structure.py
class WordDictionary:

    def __init__(self):
        self.text = []
        

    def addWord(self, word: str) -> None:
        self.text.append(word)
        

    def search(self, word: str) -> bool:
        if '.' not in word:
            return word in self.text

        for t in self.text:
            if len(t) != len(word):
                continue
            for i in range(len(word)):
                if word[i] == '.':
                    continue
                if word[i] != t[i]:
                    break
            else:
                return True
        return False
        


# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)